### ISSUE TITLE
Fine-tuning mechanism and difference between pretrain and textcat?

### ISSUE THREAD

**Issue by TomLin** [explosion/spaCy#3759]
Hello,
I've read through relevant issues for training custom text classifier either from scratch or using pre-train model.

* [explanation of the textcat's architecture](https://github.com/explosion/spaCy/issues/1550)
* [code example for adopting pre-train model in textcat](https://github.com/explosion/spaCy/issues/3448)

But I still  have a couple of questions and it would be great if they are clarified. 

1. I suppose that in [this example](https://spacy.io/usage/training#textcat), the textcat is trained from scratch, no transfer-learning is incorporated, am I correct? More specifically, for [all three architectures which are `bow`, `simple_cnn`, `emsemble`](https://spacy.io/api/textcategorizer#architectures), do they all use word-embedding as inputs? In addition, are these word-embedding trained on the fly, instead of using word2vec or Glove etc.? 

Even if we load model such as en_core_web_lg, the textcat will still train from scratch? For in my understanding, the en_core_web_lg contains word vectors, so I am not sure if it is used in textcat?

```python
import spacy
nlp = spacy.load('en_core_web_lg')

if 'textcat' not in nlp.pipe_names:
    textcat = nlp.create_pipe('textcat')
    nlp.add_pipe(textcat, last=True)
# otherwise, get it, so we can add labels to it
else:
    textcat = nlp.get_pipe('textcat')

other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'textcat']
with nlp.disable_pipes(*other_pipes):  # only train textcat
    optimizer = nlp.begin_training()
    print("Training the model...")
    # print('{:^5}\t{:^5}\t{:^5}\t{:^5}'.format('LOSS', 'P', 'R', 'F'))
    for i in range(n_iter):
        losses = {}
        # batch up the examples using spaCy's minibatch
        batches = minibatch(train_ls, size=compounding(4., 32., 1.001))
        for batch in batches:
            texts, annotations = zip(*batch)
            nlp.update(texts, annotations, sgd=optimizer, drop=0.2,
                       losses=losses)
```

2. Regarding using pre-train command, does it mean that we utilize the pre-trained word embedding from spacy's custom model which is a modified version from other existing pre-train model such as Glove or BERT and just fine tune its vectors based on our specific corpus/documents in this phase.

3. If so, does that mean we are still training all parameters in the neural network's hidden layers from scratch, the difference lays only on whether we use pre-trained word embedding or not. For as far I know, the BERT's fine-tuning only adjusts the final layer for classification, but doesn't fine-tune its parameters in hidden layers. I just wonder if spacy's pre-training is different from that.

4. Just to double confirm, if we're going to use pre-train in spacy, then the textcat architecture will always need to be specified as `simple_cnn`, not the rest two architectures, am I correct?

Thanks again for your clarification, I am really appreciated.

**Comment by jfelectron** [comment]
2. Not quite. If you look at the pretraining task, it's trying to predict the vector and the loss is based on the MSE between the predicted vector and the lookup vector. Through this task,  new vectors aren't generated but the weights of the Tok2Vec encoder are fine-tuned to better match the general structure of your corpus. This is the overall goal of LM fine-tuning.

3. The language model that is involved in Spacy's pretraining is the Tok2Vec model which embeds input tokens to vectors. This is equivalent to BERT/ELMO, which has have learned a generic language encoder through their training tasks.  The Tok2Vec architecture can be found here: https://github.com/explosion/spaCy/blob/master/spacy/_ml.py#L363  . The layers of the TextCat classifier are trained from scratch. You can see that Tok2Vec is the first layer of the TextCat mode: https://github.com/explosion/spaCy/blob/master/spacy/_ml.py#L587.   Spacy pretaining is a tradeoff between transfer learning from a general language model and the overall design of Spacy, which favors lightweight models that run well on CPU. BERT et all are giant models that are hard to train on single commodity GPU. If you're chasing SOTA benchmarks, maybe it's for you, but most of us contend with performance versus ease of training and deployment. 

4. I don't think this is strictly the case but @honnibal hasn't clarified. Looking at the code, I think there are differences in layer hyperparams that conflict with the linear layer, but with proper initialization using ensemble should be possible.

**Comment by honnibal** [comment]
I understand that this is a bit confusing; I've had trouble explaining things clearly because the word vectors and the "`spacy pretrain`" are two similar but slightly different ways of doing transfer learning. There's also terminological confusing in the term "word vector", which can refer to a row in an embeddings table, or the vector assigned to some token.

For temporary clarity, let's call a model like the one produced by word2vec, GloVe etc, a "static embeddings table". It's a matrix of shape `(M, D)`, where `D` is the number of dimensions. We can map assign vectors to words using the embeddings table if we have a dictionary that maps the words to rows. The `en_core_web_md`, `en_core_web_lg` and `en_vectors_web_lg` models all provide static embeddings tables.

spaCy's parser, tagger, NER and textcat models can optionally use static embeddings tables as features. Additionally, the models always learn vector representations for the words as well. So there's always a learned component to the word representations --- it's just that sometimes, there's *also* a static component, used as an extra feature. But we never backprop to this static component.

spaCy's  parser, tagger, NER and textcat models also refine the token representations using the surrounding context. Before training begins, the trainable weights for the `tok2vec` part of the model are usually initialized randomly. If static embedding tables are being used, they of course won't be initialized randomly -- those aren't learnable parameters of the model.

Instead of initializing randomly, we can use `spacy pretrain` to start off with a better representation. To use `spacy pretrain`, we have to be pretraining the same network architecture and we expect to use at runtime. So if we want to use a static embedding table as features at runtime, we have to use this during pretraining as well.

A point of potential confusion is that the `spacy pretrain` process uses a static embedding table as part of the *objective*, regardless of whether you're using it as an input feature. So you always need to provide a static embedding table to `spacy pretrain`...It's just that sometimes you'll also use it as an input feature (if you want that at runtime), and sometimes you won't. Usually using the static embedding table as a feature gets you better accuracy, especially if your training data is small. It does make the model bigger though, and it's slower to load.

> Just to double confirm, if we're going to use pre-train in spacy, then the textcat architecture will always need to be specified as simple_cnn, not the rest two architectures, am I correct?

That's currently true, yes. I hope to relax this restriction in future.
